app/Platform/Actions/ResetPlayerProgress.php

```diff
@@ -47,13 +49,20 @@ public function execute(User $user, ?int $achievementID = null, ?int $gameID = n
         if ($achievementID !== null) {
             $playerAchievement = $user->playerAchievements()->where('achievement_id', $achievementID)->first();
             $achievement = $playerAchievement->achievement;
-            if ($playerAchievement->unlocked_hardcore_at && $achievement->isPublished) {
-                // resetting a hardcore unlock removes hardcore mastery badges
-                $user->playerBadges()
+            if ($achievement->isPublished) {
+                // resetting a published achievement removes the completion/mastery badge.
+                // RevalidateAchievementSetBadgeEligibility will be called indirectly from
+                // the UpdatePlayerGameMetricsJob, but it does not revoke badges unless all
+                // achievements for a game are reset.
+                $playerBadge = $user->playerBadges()
                     ->where('AwardType', AwardType::Mastery)
                     ->where('AwardData', $achievement->game_id)
-                    ->where('AwardDataExtra', UnlockMode::Hardcore)
-                    ->delete();
+                    ->where('AwardDataExtra', $playerAchievement->unlocked_hardcore_at ? UnlockMode::Hardcore : UnlockMode::Softcore)
+                    ->first();
+                if ($playerBadge) {
+                    PlayerBadgeLost::dispatch($playerBadge);
+                    $playerBadge->delete();
```

Comment:
Event should follow the action to keep things in order

Reply:
This is following the logic in RevalidateAchievementSetBadgeEligibility: https://github.com/RetroAchievements/RAWeb/blob/c92999b87927ef4c0b2922720f6232f3e85317d6/app/Platform/Actions/RevalidateAchievementSetBadgeEligibility.php#L103-L104

Though there seems to be a problem with the lifetime of the badge even if it's deleted after raising the event: https://discord.com/channels/476211979464343552/1026595325038833725/1173661914556006431